File: Framework.py
#Framework program using python

fields_file = "FieldNames.cfg"
menu_file = "Menu.cfg"
data_file = "Data.dat"
updatable_fields_file = "UpdatableFields.cfg"


def add_record():
	status = True
	field_values = []
	field_values.append(status)
	for field_name in field_names:
		print("Enter", field_name, ':', end = "")
		field_value = input()
		field_values.append(field_value)
	records.append(field_values)
	write_data_into_file(records)
	print("Record inserted successfully.")

def search_record():
	id = input("Enter " + field_names[0] + " to find: ")
	is_record_found = False
	for record in records:
		if record[0] == True and record[1] == id:
			is_record_found = True
			print_record(record)
	check_record_found_or_not(is_record_found)

def check_record_found_or_not(record_status):
	if record_status == False:
		print("Record not found.")

def print_all_records():
	for record in records:
		if record[0] == True:
			print("*******************")
			print_record(record)
			
def print_record(record):
	index = 1
	for field_name in field_names:
				print(field_name + ':', record[index])
				index += 1

def update_record():
	id = input("Enter " + field_names[0] + " to find: ")
	updatable_fields_file_obj = open(updatable_fields_file)
	updatable_fields = updatable_fields_file_obj.read()
	updatable_fields_file_obj.close()
	is_record_found = False
	for record in records:
		if record[0] == True and record[1] == id:
			is_record_found = True
			counter = 1
			for updatable_field_position in range(len(updatable_fields)):
				print(str(counter) + '. ' + field_names[int(updatable_fields[counter - 1]) - 1])
				counter += 1
			update_option = int(input("Choose your option: "))
			print("Enter new " + field_names[int(updatable_fields[update_option - 1]) - 1] + ": ", end = "")
			record[int(updatable_fields[update_option - 1])] = input()
			write_data_into_file(records)
			print("Record updated successfully.")
	check_record_found_or_not(is_record_found)
def delete_record():
	id = input("Enter " + field_names[0] + " to find: ")
	is_record_found = False
	for record in records:
		if record[0] == True and record[1] == id:
			is_record_found = True
			record[0] = False
			write_data_into_file(records)
			print("Record deleted successfully.")
	check_record_found_or_not(is_record_found)

def write_data_into_file(records):
	data_file_obj = open(data_file, 'w')
	data_file_obj.write(str(records))
	data_file_obj.close()

def is_file_present(file_name):
	try:
		file_obj = open(file_name)
	except Exception:
		print("File not found or error in opening.")
		exit()

field_names = []
is_file_present(fields_file)
field_names_obj = open(fields_file)
for field_name in field_names_obj.readlines():
	field_name = field_name.rstrip()
	field_names.append(field_name)
field_names_obj.close()

is_file_present(menu_file)
menu_file_obj = open(menu_file)
menu_data = menu_file_obj.read()
menu_file_obj.close()

is_file_present(data_file)
data_file_obj = open(data_file)
records = data_file_obj.read()
data_file_obj.close()
records = eval(records)

while True:
	print(menu_data)
	functions = [add_record, search_record, print_all_records, update_record, delete_record, exit]
	user_option = int(input("Choose your option: "))
	if user_option == 6:
		print("Are you sure? ")
		print("1. Exit\n2. Cancel\nChoose your option: ", end = "")
		exit_option = int(input())
		if exit_option == 1:
			exit()
		else:
			continue
	elif user_option > 0 and user_option < 6:
		functions[user_option - 1]()
	else:
		print("INVALID OPTION.")

File: framework.java
// Framework program in Java

import java.sql.*;
import java.util.Scanner;

class MainClass
{
	public static void main(String args[])
	{
		Framework obj = new Framework();
		try
		{
			obj.printMenu();
		}
		catch(Exception e)
		{
			System.out.println(e.getMessage());
		}

	}
}

class Framework
{
	Scanner scan = new Scanner(System.in);
	Statement stmt;
	ResultSet result;
	Connection conn;
	String[] fieldNames;
	String[] promptMessages;
	static String errorMessage = "Invalid, please enter a valid option.";

	public Framework()
	{
		String url = "jdbc:mysql://165.22.14.77/dbSheshu?user=Sheshu&password=Sheshu";
		try
		{
			conn = DriverManager.getConnection(url);
			System.out.println("Connection successful.");
			stmt = conn.createStatement();
			storeFieldNames();
			String messages = getData("Messages");
			promptMessages = messages.split(", ");
		}
		catch(Exception e)
		{
			System.out.println(e.getMessage());
		}
	}

	public void printMenu() throws SQLException
	{
		String menu = getData("Menu");
		while(true)
		{
			System.out.print(menu + "\nChoose your option: ");
			int userOption = scan.nextInt();
			switch(userOption)
			{
				case 1:
					addRecord();
					break;
				case 2:
					searchRecord();
					break;
				case 3:
					printRecords();
					break;
				case 4:
					updateRecord();
					break;
				case 5:
					deleteRecord();
					break;
				case 6:
					System.out.println("Do you really want to exit? ");
					System.out.println("Enter 'Y' to exit or 'N' to continue: ");
					String exitOption = scan.next();
					switch(exitOption.toUpperCase())
					{
						case "Y":
							stmt.close();
							conn.close();
							System.exit(0);
						case "N":
							continue;
						default:
							System.out.println(errorMessage);
					}
					break;
				default:
					System.out.println(errorMessage);
			}
		}
	}

	private String getData(String fileName) throws SQLException
	{
		result = stmt.executeQuery("select Content from Config where filename = '" + fileName + "'");
		String data = "";
		while(result.next())
		{
			data = result.getString("Content");
		}
		return data;
	}

	private void storeFieldNames() throws SQLException
	{
		result = stmt.executeQuery("Select * from my_table");
		ResultSetMetaData rsmd = result.getMetaData();
		fieldNames = new String[rsmd.getColumnCount() - 1];
		for(int index = 0; index < fieldNames.length; index++)
		{
			if(rsmd.getColumnName(index + 1).equals("Status") != true)
			{
				fieldNames[index] = rsmd.getColumnName(index + 1);
			}
		}
	}

	public void addRecord() throws SQLException
	{
		String fieldValues = "";
		for(int index = 0; index < fieldNames.length; index++)
		{
			System.out.print(fieldNames[index] + ": ");
			String fieldValue = scan.nextLine();
			fieldValues += "'" + fieldValue + "', ";
		}
		fieldValues += "'A'";
		int rowsAffected = stmt.executeUpdate("insert into my_table values(" + fieldValues + ")");
		if(rowsAffected == 1)
		{
			System.out.println(promptMessages[0]);
		}
		
	}

	public void searchRecord() throws SQLException
	{
		String recordToBeSearched = getInput();
		boolean isRecordFound = checkWhetherRecordPresentOrNot(recordToBeSearched);
		if(isRecordFound)
		{
			result = stmt.executeQuery("select * from my_table where Status = 'A' and " + fieldNames[0] + " = " + recordToBeSearched);
			printRecord(result);
		}
		else
		{
			printRecordNotFound();
		}
	}

	private void printRecord(ResultSet result) throws SQLException
	{
		while(result.next())
		{
			System.out.println("-----------------------");
			for(String fieldName : fieldNames)
			{
				System.out.println(fieldName + ": " + result.getString(fieldName));
			}
		}
	}

	public void printRecords() throws SQLException
	{
		result = stmt.executeQuery("select * from my_table where Status = 'A'");
		printRecord(result);	
	}

	public void updateRecord() throws SQLException
	{
		String recordToBeUpdated = getInput();
		boolean isRecordFound = checkWhetherRecordPresentOrNot(recordToBeUpdated);
		if(isRecordFound)
		{
			String tempUpdatableFields = getData("UpdatableFields");
			String updatableFields[] = tempUpdatableFields.split(",");
			int[] updatableFieldPositions = new int[updatableFields.length];
			for(int index = 0; index < updatableFieldPositions.length; index++)
			{
				updatableFieldPositions[index] = Integer.parseInt(updatableFields[index]);
			}
			int counter = 1;
			for(int index = 0; index < updatableFieldPositions.length; index++)
			{
				System.out.println(counter + ". Update " + fieldNames[updatableFieldPositions[index] - 1]);
				counter++;
			}
			System.out.print("Choose your option: ");
			int updateOption = scan.nextInt();
			System.out.print("Enter new " + fieldNames[updatableFieldPositions[updateOption - 1] - 1] + ": ");
			String updateValue = scan.next();
			int rowsAffected = stmt.executeUpdate("update my_table set " + fieldNames[updatableFieldPositions[updateOption - 1] - 1] + " = '" + updateValue + "' where " + fieldNames[0] + " = " + recordToBeUpdated);
			if(rowsAffected == 1)
			{
				System.out.println(promptMessages[1]);
			}
		}
		else
		{
			printRecordNotFound();
		}
	}

	public void deleteRecord() throws SQLException
	{
		String recordToBeDeleted = getInput();
		boolean isRecordFound = checkWhetherRecordPresentOrNot(recordToBeDeleted);
		if(isRecordFound)
		{
			int rowsAffected = stmt.executeUpdate("update my_table set Status = 'D' where " + fieldNames[0] + " = " + recordToBeDeleted);
			if(rowsAffected == 1)
			{
				System.out.println(promptMessages[2]);
			}
		}
		else
		{
			printRecordNotFound();
		}
	}

	private void printRecordNotFound()
	{
		System.out.println(promptMessages[3]);
	}

	private String getInput()
	{
		System.out.print("Enter " + fieldNames[0] + ": ");
		String id = scan.next();
		return id;
	}

	private boolean checkWhetherRecordPresentOrNot(String id) throws SQLException
	{
		boolean isRecordFound = false;
		result = stmt.executeQuery("select " + fieldNames[0] + " from my_table where " + fieldNames[0] + " = " + id);
		if(result.next())
		{
			return isRecordFound = true;
		}
		return isRecordFound;
	}
}